php-frontend/src/main/java/org/sonar/php/highlighter/HighlighterVisitor.java

```diff
@@ -0,0 +1,130 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.php.highlighter;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.php.api.PHPKeyword;
+import org.sonar.php.tree.impl.expression.LiteralTreeImpl;
+import org.sonar.php.tree.impl.lexical.InternalSyntaxToken;
+import org.sonar.plugins.php.api.tree.Tree;
+import org.sonar.plugins.php.api.tree.Tree.Kind;
+import org.sonar.plugins.php.api.tree.lexical.SyntaxToken;
+import org.sonar.plugins.php.api.tree.lexical.SyntaxTrivia;
+import org.sonar.plugins.php.api.visitors.PHPSubscriptionCheck;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class HighlighterVisitor extends PHPSubscriptionCheck {
+
+  /**
+   * An array containing reserved variables.
+   */
+  private static final ImmutableList<String> PHP_RESERVED_VARIABLES = ImmutableList.of(
+    "__FUNCTION__",
+    "__CLASS__",
+    "__METHOD__",
+    "__NAMESPACE__",
+    "__DIR__",
+    "__FILE__",
+    "__LINE__",
+    "$this");
+  private static final ImmutableList<String> WORDS_TO_HIGHLIGHT = ImmutableList.<String>builder()
+    .add(PHPKeyword.getKeywordValues())
+    .addAll(PHP_RESERVED_VARIABLES).build();
+  private final List<HighlightingData> datas;
+  private final SourceFileOffsets offsets;
+
+  private HighlighterVisitor(SourceFileOffsets offsets) {
+    this.datas = new ArrayList<>();
+    this.offsets = offsets;
+  }
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(
+      Kind.NUMERIC_LITERAL,
+      Kind.REGULAR_STRING_LITERAL,
+      Kind.EXPANDABLE_STRING_LITERAL,
+      Kind.TOKEN
+    );
+  }
+
+  public List<HighlightingData> datas() {
+    return datas;
+  }
+
+  public static List<HighlightingData> getHighlightData(Tree tree, SourceFileOffsets sourceFileOffsets) {
+    HighlighterVisitor visitor = new HighlighterVisitor(sourceFileOffsets);
+    visitor.scanTree(tree);
+    return visitor.datas();
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    switch (tree.getKind()) {
+      case TOKEN:
+        highlightToken((InternalSyntaxToken) tree);
+        break;
+      case REGULAR_STRING_LITERAL:
+        SyntaxToken token = ((LiteralTreeImpl) tree).token();
+        highlight(offsets.startOffset(token), offsets.endOffset(token), "s");
+        break;
+      case NUMERIC_LITERAL:
+        token = ((LiteralTreeImpl) tree).token();
+        highlight(offsets.startOffset(token), offsets.endOffset(token), "c");
+        break;
+      case EXPANDABLE_STRING_LITERAL:
+        // Fixme nothing yet
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AS1134)